src/xml_parser_main_metadata_together.c

```diff
@@ -336,133 +328,215 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
             }
         }
 
-        while ((cbdata.in_progress_count_other <= cbdata.in_progress_count_filelists ||
-                cbdata.in_progress_count_other <= cbdata.in_progress_count_primary) &&
-               !other_is_done)
+        while ((cbdata->in_progress_count_other <= cbdata->in_progress_count_filelists ||
+                cbdata->in_progress_count_other <= cbdata->in_progress_count_primary) &&
+               !iter->other_is_done)
         {
-            other_is_done = parse_next_section(other_f, other_path, other_pd, err);
+            iter->other_is_done = parse_next_section(iter->other_f, iter->other_path, iter->other_pd, err);
             if (*err) {
-                ret = (*err)->code;
-                goto out;
+                return NULL;
             }
         }
     }
 
-out:
-    if (ret != CRE_OK) {
+    return g_queue_pop_head(cbdata->finished_pkgs_queue);
+}
+
+gboolean cr_PkgIterator_is_finished(cr_PkgIterator *iter) {
+    return iter->primary_is_done && iter->filelists_is_done && iter->other_is_done;
+}
+
+void cr_PkgIterator_free(cr_PkgIterator *iter, GError **err) {
+    assert(iter);
+
+    if (iter->tmp_err) {
         // An error already encountered
         // just close the file without error checking
-        cr_close(primary_f, NULL);
-        cr_close(filelists_f, NULL);
-        cr_close(other_f, NULL);
+        cr_close(iter->primary_f, NULL);
+        cr_close(iter->filelists_f, NULL);
+        cr_close(iter->other_f, NULL);
     } else {
+        GError* tmp_err = iter->tmp_err;
         // No error encountered yet
-        cr_close(primary_f, &tmp_err);
-        if (!tmp_err)
-            cr_close(filelists_f, &tmp_err);
-        if (!tmp_err)
-            cr_close(other_f, &tmp_err);
+        cr_close(iter->primary_f, &tmp_err);
+        if (!tmp_err) {
+            cr_close(iter->filelists_f, &tmp_err);
+        }
+        if (!tmp_err) {
+            cr_close(iter->other_f, &tmp_err);
+        }
         if (tmp_err) {
-            ret = tmp_err->code;
             g_propagate_prefixed_error(err, tmp_err, "Error while closing: ");
         }
     }
 
-    cr_xml_parser_data_free(primary_pd);
-    cr_xml_parser_data_free(filelists_pd);
-    cr_xml_parser_data_free(other_pd);
+    cr_xml_parser_data_free(iter->primary_pd);
+    cr_xml_parser_data_free(iter->filelists_pd);
+    cr_xml_parser_data_free(iter->other_pd);
 
-    if (allow_out_of_order) {
-        g_hash_table_destroy(cbdata.in_progress_pkgs_hash);
+    cr_CbData *cbdata = iter->cbdata;  // avoid dereferencing void *
+    if (iter->allow_out_of_order) {
+        g_hash_table_destroy(cbdata->in_progress_pkgs_hash);
     } else {
-        cr_slist_free_full(cbdata.in_progress_pkgs_list, (GDestroyNotify) cr_package_free);
+        cr_slist_free_full(cbdata->in_progress_pkgs_list, (GDestroyNotify) cr_package_free);
     }
 
-    return ret;
+    g_queue_free_full(cbdata->finished_pkgs_queue, (GDestroyNotify) cr_package_free);
```

Comment:
The current problems - I think - are caused by a pseudo joint-ownership of the underlying package objects between `py_pkgs` and `finished_pkgs_queue`.  

The outer Python newpkgcb creates the package object and puts it into `py_pkgs`, and simultaneously they end up on `finished_pkgs_queue`.  When an error is hit during parsing `cr_parse_main_metadata_together()` cleans up the package iterator including `finished_pkgs_queue`, but then the Python wrapper tries to clean up `py_pkgs`, which deallocates all of the packages within, except the underlying cr_package has already been freed.

But I'm not 100% sure because at this point my brain is fried. Any ideas about how to resolve that one?